### PR TITLE
add feature: azurefile mount on windows node

### PR DESCRIPTION
**Pull request by andyzhangx**
**What this PR does / why we need it**:
feature: azurefile mount on windows node. I created this new PR, close the original one(https://github.com/kubernetes/kubernetes/pull/50233) as there is a big rebase change.
Currently only SMB(a nfs protocol) is supported for windows container in the new Windows 2016 RS3 image, and windows container in RS3 could only use New-SmbGlobalMapping cmdlet for volume mapping, "net use" command does not work for windows container.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:
As there is a known blocking issue in Windows 2016 server when mounting a SMB(a NFS protocol in Windows) share on a container host and then bind that share to a container ( Azure file on Windows is using SMB protocol), this PR still could not mount an azure file on current windows 2016 server node, it depends on 2016 RS3 release, and it will still succeed (as a workaround) if customer want to mount an azure file on current windows node.

Main code logic is similar to what it does in Linux node:

1. create target directory in Windows host
2. Use New-SmbGlobalMapping powershell cmdlet to mount SMB azure file to a drive in Windows host
3. Use mklink command to link target directory to the mounted drive

K8s would bind target directory to the container directory
source in mount function would be like:
`\\[accountname].file.core.windows.net\test`

target in mount function would be like:
`c:\var\lib\kubelet\pods\5f679f75-7ce3-11e7-b718-000d3a31dac4\volumes\kubernetes.io~azure-file`

sample azure file config file:
```
apiVersion: v1
kind: Pod
metadata:
 name: iis
spec:
 containers:
  - image: microsoft/iis
    name: iis
    volumeMounts:
      - name: azure
        mountPath: "d:"
 nodeSelector:
   beta.kubernetes.io/os: windows
 volumes:
      - name: azure
        azureFile:
          secretName: azure-secret
          shareName: k8stest
          readOnly: false
```

**Release note**:

```release-note
```

### CODE REVIEW

**Comment by andyzhangx** [comment]
@rootfs @brendanburns @karataliu @JiangtianLi for code review

File: pkg/util/mount/mount_windows.go [hunk 1]
```diff
@@ -74,8 +74,23 @@ func (mounter *Mounter) Mount(source string, target string, fstype string, optio
 			return nil
 		}
 
-		// empty implementation for mounting azure file
-		return os.MkdirAll(target, 0755)
+		cmd := fmt.Sprintf(`$User = "AZURE\%s";$PWord = ConvertTo-SecureString -String "%s" -AsPlainText -Force;`+
```
Comment (rootfs) [inline pkg/util/mount/mount_windows.go]: what if it is an nfs mount?

File: pkg/util/mount/mount_windows.go [hunk 2]
```diff
@@ -74,8 +74,23 @@ func (mounter *Mounter) Mount(source string, target string, fstype string, optio
...
+
+		if output, err := exec.Command("powershell", "/c", cmd).CombinedOutput(); err != nil {
+			// we don't return error here, even though New-SmbGlobalMapping failed, we still make it successful,
+			// will return error when Windows 2016 RS3 is ready on azure
+			glog.Errorf("windowsMount: SmbGlobalMapping failed: %v, output: %q", err, string(output))
+			return os.MkdirAll(target, 0755)
+		}
 	}
 
 	if output, err := exec.Command("cmd", "/c", "mklink", "/D", target, bindSource).CombinedOutput(); err != nil {
```
Comment (rootfs) [inline pkg/util/mount/mount_windows.go]: want to make sure,  `cmd` here is a windows command, not the `cmd` variable defined above?

File: pkg/util/mount/mount_windows.go (hunk 1 above)
Comment (rootfs) [inline pkg/util/mount/mount_windows.go]: is `User="AZURE\%s` too restrictive? what if the node is not in azure cloud.

Comment (andyzhangx) [reply pkg/util/mount/mount_windows.go]: currently only SMB(a nfs protocol) is supported for windows container in the new Windows 2016 RS3 image. It's only by New-SmbGlobalMapping cmdlet that windows container could use for volume mapping, "net use" command does not work for windows container.

Comment (andyzhangx) [reply pkg/util/mount/mount_windows.go]: good point, I will put User="AZURE\%s" into azure_file.go which would be more reasonable

File: pkg/util/mount/mount_windows.go (hunk 2 above)
Comment (andyzhangx) [reply pkg/util/mount/mount_windows.go]: yes, and this "cmd" var name would be error prone, I have changed the name to cmdLine

File: pkg/util/mount/mount_windows.go [hunk 3]
```diff
@@ -74,8 +74,23 @@ func (mounter *Mounter) Mount(source string, target string, fstype string, optio
 			return nil
 		}
 
-		// empty implementation for mounting azure file
-		return os.MkdirAll(target, 0755)
+		cmdLine := fmt.Sprintf(`$User = "%s";$PWord = ConvertTo-SecureString -String "%s" -AsPlainText -Force;`+
```
Comment (rootfs) [inline pkg/util/mount/mount_windows.go]: please leave a note here explaining only smb mount is expected here and log a warning otherwise.

Comment (andyzhangx) [reply pkg/util/mount/mount_windows.go]: done

**Comment by rootfs** [comment]
/retest

@andyzhangx looks good, please squash

File: pkg/util/mount/mount_windows.go [hunk 4]
```diff
@@ -74,8 +74,24 @@ func (mounter *Mounter) Mount(source string, target string, fstype string, optio
 			return nil
 		}
 
-		// empty implementation for mounting azure file
-		return os.MkdirAll(target, 0755)
+		// currently only SMB mount is supported
```
Comment (rootfs) [inline pkg/util/mount/mount_windows.go]: exit out and log an error if fstype is not smb.

Comment (andyzhangx) [reply pkg/util/mount/mount_windows.go]: fixed, actually it should be "cifs":
https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/azure_file/azure_file.go#L232

**Comment by jingxu97** [comment]
approve no-issue

**Pull request merged**